### PR TITLE
dialog: focus lock

### PR DESCRIPTION
**Pull request by danethurber**
fixes: #260 

- [x] new `<FocusLock />`
- [x] `<Dialog />` uses focus lock

TODO: 

- [x] rename `<FocusLock />` to `<FocusManager />`
- [x] add `trapped` prop to activate/deactivate functionality
- [x] remove `as` prop
- [x] with no focusable children - make container focusable 
- [x] update `<Dialog />` to always use the `<FocusManager />`
- [x] `<Dialog />` is `trapped` if has `onClose` or `modal` props

### CODE REVIEW

File: packages/dialog/.storybook/config.js [hunk 1]
```diff
@@ -1,9 +1,22 @@
...
+import core from '@pluralsight/ps-design-system-core'
+import centerDecorator from '@pluralsight/ps-design-system-storybook-addon-center'
+import themeDecorator from '@pluralsight/ps-design-system-storybook-addon-theme'
+
+const paddingDecorator = storyFn => (
+  <div style={{ padding: core.layout.spacingXLarge }}>{storyFn()}</div>
+)
+
+addDecorator(paddingDecorator)
+// addDecorator(centerDecorator)
```
Comment (jaketrent) [inline packages/dialog/.storybook/config.js]: rm? and import?

File: packages/dialog/src/react/index.js [hunk 2]
```diff
@@ -51,39 +49,59 @@ const CloseButton = props => (
...
+    if (evt.keyCode === ESCAPE_KEYCODE) this.props.onClose(evt)
+  }
+
   render() {
     const { props } = this
     const dialogProps = {
       ...styles.dialog(props),
       ...(!this.props.disableCloseOnEscape &&
       typeof this.props.onClose === 'function'
-        ? { onKeyUp: handleKeyUp.bind(this, this.props) }
```
Comment (jaketrent) [inline packages/dialog/src/react/index.js]: Why change?  Dislike the style?  Perf tweak?

File: packages/focuslock/.babelrc [hunk 3]
```diff
@@ -0,0 +1,14 @@
+{
```
Comment (jaketrent) [inline packages/focuslock/.babelrc]: why create a whole new package for this?  It's only used in the Dialog.  Why not colocate?

File: packages/focuslock/src/react/index.js [hunk 4]
```diff
@@ -0,0 +1,101 @@
+import PropTypes from 'prop-types'
+import React from 'react'
+
+const FOCUSABLE_SELECTORS = [
```
Comment (jaketrent) [inline packages/focuslock/src/react/index.js]: where did this list come from?

File: packages/focuslock/src/react/index.js [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    return (
+      <Tag ref={this.bindElement} onKeyDown={this.handleKeyDown}>
+        {children}
+      </Tag>
+    )
+  }
+}
+
+FocusLock.propTypes = {
+  as: PropTypes.oneOf(['div']).isRequired,
```
Comment (jaketrent) [inline packages/focuslock/src/react/index.js]: This is a pattern I've seen in Material UI, but we've yet to use it in our components.  

Thus far, when we've wanted to supply semantics, we've passed it as a first child, something like:

```
<FocusLock>
  <div>...
```

I like that approach better in general.  I'd rather have one approach than many.  But I haven't considered long enough whether that approach even works or feels appropriate in this case.  So, we might need another approach.  Or the advantages of this approach, so far unapparent to me, could be enough the we should migrate to it.  What is your take on this subject?

File: packages/focuslock/src/react/index.js [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+  componentDidUpdate() {
+    this.focusableNodes = getFocusableChildNodes(this.el)
+  }
+
+  componentWillUnmount() {
+    const { prevFocusedElement: prev } = this
+    if (this.props.returnFocus && prev && prev.focus) prev.focus()
+  }
+
+  handleKeyDown(event) {
```
Comment (jaketrent) [inline packages/focuslock/src/react/index.js]: beautiful implementation

File: packages/focuslock/test/setup.js [hunk 7]
```diff
@@ -0,0 +1,4 @@
+import Adapter from 'enzyme-adapter-react-16'
```
Comment (jaketrent) [inline packages/focuslock/test/setup.js]: what would you think about migrating to react-testing-library?  Pros: We don't have to do version adapters.  We avoid the ginormous enzyme api.  We are forced more into blackbox testing and dom interrogation.

File: packages/focuslock/package.json [hunk 8]
```diff
@@ -17,6 +17,9 @@
     "test:updateSnapshot": "npm run test -- --updateSnapshot",
     "test:watch": "npm run test -- --watchAll"
   },
+  "publishConfig": {
```
Comment (jaketrent) [inline packages/focuslock/package.json]: fancy!

File: packages/focuslock/src/react/__specs__/index.spec.js [hunk 9]
```diff
@@ -3,15 +3,113 @@ import { mount } from 'enzyme'
...
 
+describe('FocusLock', () => {
   beforeEach(() => {
-    props = {
-      children: 'children'
-    }
+    document.activeElement.blur()
+  })
+
+  it('has defaultProps', () => {
```
Comment (jaketrent) [inline packages/focuslock/src/react/__specs__/index.spec.js]: I have never tested default props as a structure.  I would maybe rather test the outcome that the existence of a defaultProp is meant to create.  What value have you found in testing this in the past?

File: packages/focuslock/src/react/__specs__/index.spec.js [hunk 10]
```diff
@@ -3,15 +3,113 @@ import { mount } from 'enzyme'
...
+    </li>
+  </ul>
+)
 
+describe('FocusLock', () => {
   beforeEach(() => {
-    props = {
-      children: 'children'
-    }
+    document.activeElement.blur()
```
Comment (jaketrent) [inline packages/focuslock/src/react/__specs__/index.spec.js]: oh, I love this.  Meant to "not focus on anything", right?

File: packages/focuslock/src/react/__specs__/index.spec.js [hunk 11]
```diff
@@ -3,15 +3,113 @@ import { mount } from 'enzyme'
...
+`)
+  })
+
+  describe('WITH focusable children', () => {
+    it('renders successfully', () => {
+      expect(() => mount(<FocusLock>{children}</FocusLock>)).not.toThrow()
+    })
+  })
+
+  describe('WITHOUT focusable children', () => {
```
Comment (jaketrent) [inline packages/focuslock/src/react/__specs__/index.spec.js]: I love these test cases.  In other components, I may have put these in snapshot test, generated from storybook.  But on thinking about what you're asserting here, that different children scenarios are *possible* without some child state to check against probably fits well into this more traditional unit test.

File: packages/focuslock/src/react/__specs__/index.spec.js [hunk 12]
```diff
@@ -3,15 +3,113 @@ import { mount } from 'enzyme'
...
+
+      const wrapper = mount(<FocusLock>{children}</FocusLock>)
+      const lastEl = wrapper.find('a[href]').last()
+
+      lastEl.getDOMNode().focus()
+      wrapper.unmount()
+
+      expect(document.activeElement).toEqual(button)
+    })
+  })
```
Comment (jaketrent) [inline packages/focuslock/src/react/__specs__/index.spec.js]: Do we need anything that tests the limits of the "focusable elements whitelist" in the component?

File: packages/dialog/src/react/__stories__/index.story.js [hunk 13]
```diff
@@ -8,28 +9,95 @@ import * as Text from '@pluralsight/ps-design-system-text/react'
...
+    closeAction(event)
+  }
+
+  render() {
+    const { modalIsOpen } = this.state
+
+    return (
+      <div>
+        <Text.P>
+          Macaroon danish I love candy macaroon danish toffee muffin. Cotton
```
Comment (jaketrent) [inline packages/dialog/src/react/__stories__/index.story.js]: Mmmmm...

File: packages/dialog/src/react/index.js [hunk 14]
```diff
@@ -87,32 +90,29 @@ class Dialog extends React.Component {
...
-  }
-
   handleKeyUp(evt) {
     if (evt.keyCode === ESCAPE_KEYCODE) this.props.onClose(evt)
   }
 
   render() {
     const { props } = this
+
+    const hasFocusLock = !!props.onClose
```
Comment (jaketrent) [inline packages/dialog/src/react/index.js]: should include `|| props.modal` ? (the case where we have a modal that doesn't feature a close button/callback)

File: packages/dialog/src/react/index.js [hunk 15]
```diff
@@ -87,32 +90,29 @@ class Dialog extends React.Component {
     this.handleKeyUp = this.handleKeyUp.bind(this)
   }
 
-  componentDidMount() {
-    if (this.el && !this.props.disableFocusOnMount) this.el.focus()
```
Comment (jaketrent) [inline packages/dialog/src/react/index.js]: What was your reasoning in getting rid of this?  I can't see where its function is handled elsewhere.

File: packages/dialog/src/react/index.js [hunk 16]
```diff
@@ -60,17 +60,16 @@ class ModalOverlay extends React.Component {
...
   }
 
   render() {
-    const { disableCloseOnOverlayClick } = this.props
-
     return (
       <div
         {...styles.overlay(this.props)}
         id={MODAL_OVERLAY_ID}
-        onClick={!disableCloseOnOverlayClick && this.handleOverlayClick}
```
Comment (jaketrent) [inline packages/dialog/src/react/index.js]: I like the change.  Just interested in what the warning was that was being addressed.

File: packages/dialog/.storybook/config.js (hunk 1 above)
Comment (danethurber) [reply packages/dialog/.storybook/config.js]: i'll fix

File: packages/focuslock/.babelrc (hunk 3 above)
Comment (danethurber) [reply packages/focuslock/.babelrc]: i have plans for reuse

File: packages/focuslock/src/react/index.js (hunk 5 above)
Comment (danethurber) [reply packages/focuslock/src/react/index.js]: i will removed - we add in support for additional tags if needed later

File: packages/focuslock/src/react/__specs__/index.spec.js (hunk 12 above)
Comment (danethurber) [reply packages/focuslock/src/react/__specs__/index.spec.js]: let's leave them out

File: packages/dialog/src/react/index.js (hunk 14 above)
Comment (danethurber) [reply packages/dialog/src/react/index.js]: i will add that prop support

File: packages/dialog/src/react/index.js (hunk 15 above)
Comment (danethurber) [reply packages/dialog/src/react/index.js]: i will update the `FocusLock` to make it's self focusable if no focusable children

Comment (danethurber) [reply packages/dialog/src/react/index.js]: Dialog(that aren't modal/onClose) should also focus on mount

Comment (danethurber) [reply packages/dialog/src/react/index.js]: update FocusLock to have active/inactive or locked/unlocked. rename FocusManager? trapped(bool?)

**Comment by danethurber** [comment]
@jaketrent this is ready for you to give it a second pass

**Comment by jaketrent** [comment]
Added one relevant story.  Please have a look.  Overall, looks great to me, plz merge.

**Comment by jaketrent** [comment]
Now that I've committed, I'm not sure I have an approve/review button any more.

**Comment by danethurber** [comment]
that's weird

**Pull request merged**